### ISSUE TITLE
Implement draft-ietf-httpbis-message-signatures-04

### ISSUE THREAD

**Issue by aljones15** [digitalbazaar/http-signature-header#45]
1. See related issue on [supporting structured fields](https://github.com/digitalbazaar/http-signature-header/issues/44).
2. ensure we support `@signature-params`
3. ensure `@created` is required now
4. ensure `keyid` is lower case
5. require `alg` and change to use `alg` over `algorithm`
6. add key type check to library using `alg`
7. ensure content-identifiers are lower case
8. If content id contains `sf` use structured fields
9. add tests for all `@request-target` examples
10. ensure `@created` & `@expires` are sf-integer
11. ensure signature input is US-ASCII
12. update covered-content serialization to use `\\n`
13. add structured fields check for dictionary keys & prefixes
14. use the new Signature-Input header

**Comment by msporny** [comment]
> Are we still using hs2019 as the alg?

No, but we need one for ed25519... let's create one called `ed25519-2021` -- we'll have to lobby to get this added to the official specification.

> What happens if I have multiple signatures in Signature-Input & one of them throws before verification?

By default, if one signature fails, the call to verify should fail. We may want to provide an option where you get back a list of all signatures that succeeded/failed.

> Do I return the request with say 2/3 signatures filled?

Always design security systems to fail closed by default. So, by default, you say the signature verification for what was passed in failed. If we provide details in the error code/response, we should state which signatures failed and for what reason like: "unknown signature suite", "unknown keyid", "could not dereference keyid", etc.

> What happens if I have multiple signatures in Signature & only one of them fails verification?

Fail closed. If one signature fails, the verification function should return false.

**Comment by gausnes** [comment]
> No, but we need one for ed25519... let's create one called ed25519-2021 -- we'll have to lobby to get this added to the official specification.

It looks like `ed25519` was added to the latest version of the [draft](https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-message-signatures-09#section-6.1.2) as an official enumeration.

> By default, if one signature fails, the call to verify should fail. We may want to provide an option where you get back a list of all signatures that succeeded/failed.

I don't think this captures the specification intention of multiple signature support. I think that the invoker should specify the signature identifier they are wishing to inspect. If you inspect the [example](https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-message-signatures-09#section-4.3) in the latest draft, you can see the proxy need to override the authority. This would cause the proxied request to no longer have the context to validate `sig1`.
